File: elixir/lib/forwardfinancing.ex
defmodule ForwardFinancing do
  def api_example do

    # Read the API key from an environment variable
    api_key = System.get_env("FORWARD_FINANCING_API_KEY")

    json_data = %{
      lead: %{
        contacts_attributes: [
          %{
            "first_name" => "string",
            "last_name" => "string",
            "email" => "test@forwardfinancing.com",
            "title" => "string",
            "born_on" => "2015-01-01",
            "home_phone" => "6176781000",
            "cell_phone" => "6176781000",
            "ssn" => "234345566",
            "ownership_date" => "2015-01-01",
            current_address_attributes:  %{
              "street1" => "string",
              "street2" => "string",
              "city" => "string",
              "state" => "AK",
              "zip" => "00112"
            }
          }
        ],
        account_attributes: %{
          "entity_type" => "Sole Proprietor",
          "name" => "string",
          "started_on" => "2015-01-01",
          "legal_name" => "string",
          "phone" => "6176781000",
          "email" => "test@forwardfinancing.com",
          "website" => "string",
          "fein" => "string",
          "monthly_revenue" => "Less than $5,000",
          "industry_name" => "Laundry and dry cleaning services",
          current_address_attributes: %{
            "street1" => "string",
            "street2" => "string",
            "city" => "string",
            "state" => "AK",
            "zip" => "00112"
          }
        },
        loan_attributes:  %{
          "company_name" => "string",
          "daily_payment_amount" => 0,
          "balance" => 0
        },
        application_attributes: %{
          "has_current_loan" => true,
          "applicant_is_owner" => true,
          "loan_use" => "Debt Refinancing",
          "capital_needed" => "string",
          "owner_1_percent_ownership" => 0,
          "owner_2_percent_ownership" => 0,
          "reference_id" => "string"
        }
      }
    }

    # Encode the map in json
    json = json_data |> Poison.encode!

    # Send a lead
    response = HTTPoison.post!(
      "https://api-staging.forwardfinancing.com/v1/lead",
      headers: %{
        "api_key" => api_key,
        "Content-Type" => "application/json"
      },
      body: json
    )

    if response.status_code == 201 do
      IO.puts response.body
      lead_id = response.body |> Poison.decode! |> Map.get("id")
      IO.puts "Lead id: #{lead_id}"
    else
      # Something went wrong
      IO.puts response.status_code
    end


    # Send an attachment for the lead
    file_binary = File.read!("forwardfinancing.ex")

    response = HTTPoison.post!(
      "https://api-staging.forwardfinancing.com/v1/attachment?lead_id=#{lead_id}&filename=test.txt",
      headers: %{
        "api_key" => api_key
      },
      body: file_binary
    )

    if response.status_code == 202 do
      IO.puts "It worked!"
    end


    # Send an encoded attachment
    file_string = file_binary |> Base.encode64
    response = HTTPoison.post!(
      "https://api-staging.forwardfinancing.com/v1/attachment?lead_id=#{lead_id}&filename=test.txt&encoded=true",
      headers: %{
        "api_key" => api_key
      },
      body: file_string
    )

    if response.status_code == 202 do
      IO.puts "It worked!"
    end
  end
end

File: python3/forwardfinancing.py
import requests
import os
import json
import base64

# Since you don't want to hard-code an api key, read it from environment var
api_key = os.environ["FORWARD_FINANCING_API_KEY"]

# sample payload
body = {
  "lead": {
    "contacts_attributes": [
      {
        "first_name": "string",
        "last_name": "string",
        "email": "test@forwardfinancing.com",
        "title": "string",
        "born_on": "2015-01-01",
        "home_phone": "6176781000",
        "cell_phone": "6176781000",
        "ssn": "234345566",
        "ownership_date": "2015-01-01",
        "current_address_attributes": {
          "street1": "string",
          "street2": "string",
          "city": "string",
          "state": "AK",
          "zip": "00112"
        }
      }
    ],
    "account_attributes": {
      "entity_type": "Sole Proprietor",
      "name": "string",
      "started_on": "2015-01-01",
      "legal_name": "string",
      "phone": "6176781000",
      "email": "test@forwardfinancing.com",
      "website": "string",
      "fein": "string",
      "monthly_revenue": "Less than $5,000",
      "industry_name": "Laundry and dry cleaning services",
      "current_address_attributes": {
        "street1": "string",
        "street2": "string",
        "city": "string",
        "state": "AK",
        "zip": "00112"
      }
    },
    "loan_attributes": {
      "company_name": "string",
      "daily_payment_amount": 0,
      "balance": 0
    },
    "application_attributes": {
      "has_current_loan": True,
      "applicant_is_owner": True,
      "loan_use": "Debt Refinancing",
      "capital_needed": "10000",
      "owner_1_percent_ownership": 0,
      "owner_2_percent_ownership": 0,
      "reference_id": "unique_string_here"
    }
  }
}

# Send your api_key in in the header
# Make sure to specify the Content-Type for the /lead endpoint!
headers = {
    "api_key": api_key,
    "Content-Type": "application/json"
}

print(json.dumps(body, indent=4))

response = requests.post("https://api-staging.forwardfinancing.com/v1/lead", headers=headers, data=json.dumps(body))

# If the response was 201 it worked
if response.status_code == 201:
    print(response.json())
    lead_id = response.json()["id"]
    print("Lead id: ", lead_id)
else:
    # If the response was not 200, something went wrong
    # The response body might have some info about what wrong in addition to the
    # status code
    print(response.status_code, " - Error")

# Send an attachment
with open("forwardfinancing.py", "rb") as file_binary:
    response = requests.post("https://api-staging.forwardfinancing.com/v1/attachment?lead_id={0}&filename=test.txt".format(lead_id),
        headers={"api_key": api_key},
        data=file_binary
    )
    print(response.json())

    if response.status_code == 202:
        print("It worked!")

with open("forwardfinancing.py", "rb") as file_binary:
    # Send an attachment base64 encoded
    file_string = base64.b64encode(file_binary.read())
    
    # Send the same request, only with the encoded=true url param and the encoded
    # body
    response = requests.post("https://api-staging.forwardfinancing.com/v1/attachment?lead_id={0}&filename=test.txt&encoded=true".format(lead_id),
        headers={"api_key": api_key},
        data=file_string
    )

    print(response.json())

    if response.status_code == 202:
        print("It worked!")
